ribs/archives/_archive_base.py

```diff
@@ -251,6 +263,72 @@ def _stats_reset(self):
         self._stats = ArchiveStats(0, self.dtype(0.0), self.dtype(0.0), None,
                                    None)
 
+    def _sum_geometric_series(self, a, r, n):
+        """Compute the sum of a geometric series."""
+        if r == 1.0:
+            return a
+        return (a * (1 - pow(r, n))) / (1 - r)
+
+    def _compute_new_thresholds(self, threshold_arr, objective_batch,
+                                index_batch, learning_rate):
+        """Update thresholds.
+
+        Args:
+            threshold_arr (np.ndarray): The threshold of the cells before
+                updating. 1D array.
+            objective_batch (np.ndarray): The objective values of the solution
+                that is inserted into the archive for each cell. 1D array.
+            index_batch (np.ndarray): The archive index of the elements in
+                objective batch.
+        Returns:
+            ``nd.array`` of new thresholds where ``new_threshold_batch[i]`` is
+            the new threshold value at cell ``threshold_update_indices[i]``.
+        Raises:
+            ValueError: ``threshold_arr`` or ``objective_batch`` or
+            ``index_batch`` is empty.
+        """
+        if threshold_arr.size == 0 or objective_batch.size == 0 or index_batch.size == 0:
+            raise ValueError(
+                "Cannot compute new thresholds when input array is empty")
+
+        # Compute the number of objectives inserted into each cell.
+        objective_sizes = aggregate(index_batch,
+                                    objective_batch,
+                                    func="len",
+                                    fill_value=0)
+
+        threshold_update_indices = np.unique(np.where(objective_sizes != 0))
+
+        # Compute the sum of the objectives inserted into each cell.
+        objective_sums = aggregate(index_batch,
+                                   objective_batch,
+                                   func="sum",
+                                   fill_value=np.nan)
+
+        objective_sizes = objective_sizes[threshold_update_indices]
+        objective_sums = objective_sums[threshold_update_indices]
+
+        geometric_sums = np.array([
+            self._sum_geometric_series(1, 1 - learning_rate, k)
+            for k in objective_sizes
+        ])
+
+        a = learning_rate * objective_sums * \
+            (geometric_sums / objective_sizes)
+
+        old_threshold = np.copy(threshold_arr[threshold_update_indices])
+        old_threshold[old_threshold == -np.inf] = 0
+        b = old_threshold * np.power(
+            np.full_like(objective_sizes, 1.0 - learning_rate,
+                         dtype=self.dtype), objective_sizes)
+
+        new_threshold_batch = a + b
+        if np.any(new_threshold_batch == np.nan):
+            raise ValueError(
```

Comment:
Don't think it will happen as long as we are careful with not using np.inf in our computation. i.e. prevent users from passing in np.inf for new objective.